symfit/core/fit.py

```diff
@@ -1009,169 +1010,78 @@ class HasCovarianceMatrix(TakesData):
+        except AttributeError:
+            # Some models do not have an eval_jacobian, in which case we give up
             return None
-        if any(element is None for element in self.sigma_data.values()):
-            # If one of the sigma's was explicitly set to None, we are unable
-            # to determine the covariances.
-            return np.array(
-                [[float('nan') for p in self.model.params] for p in self.model.params]
-            )
+
         try:
-            if isinstance(self.objective, LogLikelihood):
-                # Loglikelihood is a special case that needs to be considered
-                # separately, see #138
-                hess = self.objective.eval_hessian(**key2str(best_fit_params))
-                cov_mat = np.linalg.inv(hess)
-                return cov_mat
-            if len(set(arr.shape for arr in self.sigma_data.values())) == 1:
-                # Shapes of all sigma data identical
-                return self._cov_mat_equal_lenghts(best_fit_params=best_fit_params)
-            else:
-                return self._cov_mat_unequal_lenghts(best_fit_params=best_fit_params)
+            hess_inv = np.linalg.inv(hess)
         except np.linalg.LinAlgError:
             return None
 
-    def _reduced_residual_ss(self, best_fit_params, flatten=False):
-        """
-        Calculate the residual Sum of Squares divided by the d.o.f..
-
-        :param best_fit_params: ``dict`` of best fit parameters as given by .best_fit_params()
-        :param flatten: when `True`, return the total sum of squares (SS).
-            If `False`, return the componentwise SS.
-        :return: The reduced residual sum of squares.
-        """
-        # popt = [best_fit_params[p.name] for p in self.model.params]
-        # Rescale the covariance matrix with the residual variance
-        if isinstance(self.objective, (VectorLeastSquares, LeastSquares)):
-            ss_res = self.objective(flatten_components=flatten,
-                                    **key2str(best_fit_params))
-        else:
-            ss_res = self.objective(**key2str(best_fit_params))
-
-        if isinstance(self.objective, VectorLeastSquares):
-            ss_res = np.sum(ss_res**2)
-
-        degrees_of_freedom = 0 if flatten else []
-        for data in self.dependent_data.values():
-            if flatten:
-                if data is not None:
-                    degrees_of_freedom += np.product(data.shape)
+        if isinstance(objective, LogLikelihood):
+            # Loglikelihood is a special case that needs to be considered
+            # separately, see #138.
+            return hess_inv
+        elif isinstance(objective, LeastSquares):
+            # Calculate the covariance for a least squares method.
+            # https://www8.cs.umu.se/kurser/5DA001/HT07/lectures/lsq-handouts.pdf
+            # Residual sum of squares
+            rss = objective(**key2str(best_fit_params))
```

Comment:
The RSS should probably be a "goodness of fit" descriptor, and every minimizer will return it. Would save a function eval

Reply:
I agree, but maybe we should keep that for when we fix the goodness of fit mayhem ;).